draco/asp/soft.lp

```diff
@@ -30,18 +30,67 @@ preference(same_field,F) :-
     entity(mark,_,M),
     { entity(encoding,M,E): attribute((encoding,field),E,F) } = 2.
 
-% @soft(type_linear) linear scale
-preference(type_linear,E) :-
+% @soft(same_field_grt3) Prefer not to use the same field three or more times for the same mark.
+preference(same_field_grt3,F) :-
+    entity(field,_,F),
+    entity(mark,_,M),
+    { entity(encoding,M,E): attribute((encoding,field),E,F) } >= 3.
+
+% @soft(count_grt1) Prefer not to use count more than once.
+preference(count_grt1,M) :-
+    entity(mark,_,M),
+    { entity(encoding,M,E): attribute((encoding,aggregate),E,count) } > 1.
+
+% @soft(shape_high_cardinality) Shape channel should not have too high cardinality.
```

Comment:
I think you need to include that the scale is discrete(discrete data or binned). 
the `mark_channel_discrete_or_binned` helper might be relevant.

Reply:
Isn't it even number data type (continuous) can also have low cardinality, it doesn't necessarily to be discrete.

Reply:
but scale for shape need to be discrete? maybe we can let a hard constraint handle this